[docs] Add page for `experimentalStyled()` (#26361)
mui-org/material-ui: docs/src/pages/customization/styled/styled.md

# `experimentalStyled()`

<p class="description">Utility for creating styled components.</p>

## Introduction

All Material-UI components use the `experimentalStyled()` utility from `@material-ui/core/styles` under the hood.

> The utility is built on top of [emotion's `styled()` utility](https://emotion.sh/docs/styled) but provides additional options that can be used for adding some features that are specific to the Material-UI components.

Reviewer: ```suggestion
The utility is built on top of [emotion's `styled()` utility](https://emotion.sh/docs/styled) but provides additional options that can be used for adding some features that are specific to Material-UI's components.
```

The additional features include:

- support for [`theme.components[name].styleOverrides`](/customization/theme-components/#global-style-overrides)
- support for [`theme.components[name].variants`](/customization/theme-components/#adding-new-component-variants)
- support for [the `sx` prop](/system/basics/#the-sx-prop)

### What problems does it solve?

> The utility can be used as a replacement for emotion's/styled-components' styled() utility.

Reviewer: ```suggestion
The utility can be used as a replacement for emotion's or styled-components' styled() utility.
```

It aims to solve the same problem, but also provides the following benefits:

> 1. It uses the default Material-UI's `theme` if no theme is available in the context

Reviewer: Two alternatives:
```suggestion
1. It uses the default Material-UI `theme` if no theme is available in the context.
```

```suggestion
1. It uses Material-UI's default `theme` if no theme is available in the context.
```

Reviewer: A separate (additional) suggestion:
```suggestion
1. It uses the default Material-UI's `theme` if no theme is available in React context.
```
?

Author: ```suggestion
1. It uses Material-UI's default `theme` if no theme is available in React context.
```

> 2. It supports the theme's `styleOverrides` and `variants` to be applied, based on the `name` applied in the options (can be skipped)

Reviewer: ```suggestion
2. It supports the theme's `styleOverrides` and `variants` to be applied, based on the `name` applied in the options (can be skipped).
```

> 3. It adds support for the `sx` prop automatically (can be skipped)

Reviewer: ```suggestion
3. It adds support for the `sx` prop automatically (can be skipped).
```

> 4. It adds by default `shouldForwardProp` option that is taking into account all props used internally in the Material-UI components (can be overridden)

Reviewer: ```suggestion
4. It adds by default `shouldForwardProp` option that is taking into account all props used internally in the Material-UI components (can be overridden).
```

## API

> ### `experimentalStyled(Component, [options], [muiOptions])(styles) => Component`

Reviewer 2: ```suggestion
### `experimentalStyled(Component, [options])(styles) => Component`
```

#### Arguments

1. `Component`: The component that will be wrapped.
2. `options` (_object_ [optional]):

>    - `options.shouldForwardProp` (_`(props: string) => boolean`_ [optional]): Indicates whether the `prop` should be forwared to the `Component`.

Reviewer: ```suggestion
   - `options.shouldForwardProp` (_`(props: string) => boolean`_ [optional]): Indicates whether the `prop` should be forwarded to the `Component`.
```

>    - `options.shouldForwardProp` (_`(props: string) => boolean`_ [optional]): Indicates whether the `prop` should be forwarded to the `Component`.

Reviewer 3: Per #25970

```suggestion
   - `options.shouldForwardProp` (_`(props: string) => bool`_ [optional]): Indicates whether the `prop` should be forwarded to the `Component`.
```

   - `options.label` (_string_ [optional]): The suffix of the style sheet. Useful for debugging.

>    - `options.name` (_string_ [optional]): The key used under `theme.components` for specifying `styleOverrides` and `variants`. Used also for generating the `label`.

Reviewer: ```suggestion
   - `options.name` (_string_ [optional]): The key used under `theme.components` for specifying `styleOverrides` and `variants`. Also used for generating the `label`.
```

>    - `options.slot` (_string_ [optional]): If `Root`, it automatically applies the theme's `styleOverrides` & `variants`

Reviewer: ```suggestion
   - `options.slot` (_string_ [optional]): If `Root`, it automatically applies the theme's `styleOverrides` & `variants`.
```

   - `options.overridesResolver` (_(props: object, styles: Record<string, styles>) => styles_ [optional]): Function that returns styles based on the props and the `theme.components[name]styleOverrides` object.

>    - `options.skipVariantsResolver` (_boolean_): Disables the automatic resolver for the `theme.components[name].variants`

Reviewer: ```suggestion
   - `options.skipVariantsResolver` (_boolean_): Disables the automatic resolver for the `theme.components[name].variants`.
```

>    - `options.skipVariantsResolver` (_boolean_): Disables the automatic resolver for the `theme.components[name].variants`.

Reviewer 3: Per #25970

```suggestion
   - `options.skipVariantsResolver` (_bool_): Disables the automatic resolver for the `theme.components[name].variants`.
```

   - `options.skipSx` (_bool_ [optional]): Disables the `sx` prop on the component.

>    - The other keys are forwarded to the options argument of emotion's [styled([Component], [options])](https://emotion.sh/docs/styled).

Reviewer: ```suggestion
   - The other keys are forwarded to the `options` argument of emotion's [`styled([Component], [options])`](https://emotion.sh/docs/styled).
```

#### Returns

`Component`: The new component created.

## Basic usage

{{"demo": "pages/customization/styled/BasicUsage.js", "defaultCodeOpen": true}}

## Options

{{"demo": "pages/customization/styled/UsingOptions.js", "defaultCodeOpen": true, "iframe": true }}

> If you inspect this element with the dev tools, you will notice that the class of the compoent now ends with the `MyTestComponent-root`, which comes from the `name` and `slot` options that were provided. In addition to this, the `color` and `variant` props are not propagated to the generated `div` element.

Reviewer 2: ```suggestion
If you inspect this element with the browser DevTools, you will notice that the class of the component now ends with the `MyTestComponent-root`, which comes from the `name` and `slot` options that were provided. In addition to this, the `color` and `variant` props are not propagated to the generated `div` element.
```
Because we also have react devtools and the screenshot uses browser devtools.

Also used the same branding [chrome](https://developer.chrome.com/docs/devtools/), [firefox](https://twitter.com/FirefoxDevTools?ref_src=twsrc%5Egoogle%7Ctwcamp%5Eserp%7Ctwgr%5Eauthor) and [react](https://reactjs.org/blog/2019/08/15/new-react-devtools.html) use

Reviewer: "DevTools" seems to be specific to Chrome.

Reviewer 2: > "DevTools" seems to be specific to Chrome.

I'm not following how you come to this conclussion from the links:

![Screenshot from 2021-05-21 23-05-48](https://user-images.githubusercontent.com/12292047/119198140-2214b380-ba89-11eb-97ef-a99945521370.png)
![Screenshot from 2021-05-21 23-06-16](https://user-images.githubusercontent.com/12292047/119198163-293bc180-ba89-11eb-9737-b365b46d109a.png)

> <img src="/static/images/customization/styled-options.png" alt="Developer tools showing the rendered component" width="327" />

Reviewer 3: One detail, the image looks blurry. The intrinsic image width is 625px, 625/327 = 1.91. Having a ratio equals to 2 would allow it to render correctly on screen with a pixel density of 1/2/4.

Author: Changed the `width` to 312, looks better on my screen now 👍

Reviewer 3: Could we resize the image to be 624px width?

Author: You mean the `img` tag or the image itself?

Reviewer 3: I meant so that there is not sub pixel resizing. If the `<img>` width is 312px, I believe we need 624px for the size of the image.

Author: Done

# `experimentalStyled()`

> <p class="description">Utility for creating styled components, compatible with the Material-UI's core components.</p>

Reviewer 2: What are "core components"? In the introduction section we then say "all componnents" use this utility. Is there a difference between "being compatible with" (what does that mean?) and "a component is using that utility"?

Author: ```suggestion
<p class="description">Utility for creating styled components.</p>
```

This should be enough, later on, we explain how the util is connected to the Material-UI theme.

#### Arguments

1. `Component`: The component that will be wrapped.
2. `options` (_object_ [optional]):

   - `options.shouldForwardProp` (_`(props: string) => boolean`_ [optional]): Indicates whether the `prop` should be forwared to the `Component`.
   - `options.label` (_string_ [optional]): The suffix of the style sheet. Useful for debugging.
   - The other keys are forwarded to the options argument of emotion's [styled([Component], [options])](https://emotion.sh/docs/styled).

3. `muiOptions` (_object_ [optional]):

   - `muiOptions.name` (_string_ [optional]): The key used under `theme.components` for specifying `styleOverrides` and `variants`. Used also for generating the `label`.
   - `muiOptions.slot` (_string_ [optional]): If `Root`, it automatically applies the theme's `styleOverrides` & `variants`
   - `muiOptions.overridesResolver` (_(props: object, styles: Record<string, styles>) => styles_ [optional]): Function that returns styles based on the props and the `theme.components[name]styleOverrides` object.
   - `muiOptions.skipVariantsResolver` (_boolean_): Disables the automatic resolver for the `theme.components[name].variants`

>    - `muiOptions.skipSx` (_bool_ [optional]): Disables the `sx` prop on the component.

Reviewer 2: How is having two option bags relevant to the usage?

Reviewer 3: Should we have a single `options` argument instead of two?

Author: We could merge the options. Should I create a PR before proceeding with this?

Reviewer 2: Yeah, this is more of a general question now that I have the full interface+docs in front of me. So it seems like we can merge these two bags. Unless there's a meaningful difference for users for an option being a "normal" option or a "mui" option. But I suspect you wouldn't really care about whether an option is in `options` or `muiOptions`.

Author: Semantically the first option is emotion's original optional, the second are the mui specific. I don't think people will care about this, it may even be more confusing as to why we have two options, for people not aware too much with the specifics of emotion. Let me merge the options and update this PR.

#### Returns

`Component`: The new component created.

## Basic usage

{{"demo": "pages/customization/styled/BasicUsage.js", "defaultCodeOpen": true}}

## Options

{{"demo": "pages/customization/styled/UsingOptions.js", "defaultCodeOpen": true}}

If you inspect this element with the dev tools, you will notice that the class of the compoent now ends with the `MyTestComponent`, which is the label that we provided. In addition to this, the `color` prop is not propagated to the generated `div` element.

> <img src="/static/images/customization/styled-options.png" alt="dev-tools" width="327" />

Reviewer 2: The `alt` needs some work.

On the preview it's also not displayed for me.

Author: Fixed.

## `muiOptions`

{{"demo": "pages/customization/styled/UsingMuiOptions.js"}}

> If you're using TypeScript, you'll need to specify your new component's overrides/variants, using [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation).

Reviewer 2: An example would help here.

Author: I've updated the example to showcase this and removed this section.

> ### Removing some MUI feature

Reviewer 2: ```suggestion
### Removing some Material-UI feature
```

> If you would like to remove some of the MUI specific features, you can do it like this:

Reviewer 2: ```suggestion
If you would like to remove some of the Material-UI specific features, you can do it like this:
```
